File: [5] Longest Palindromic Substring/5.longest-palindromic-substring.cpp
/*
 * @lc app=leetcode id=5 lang=cpp
 *
 * [5] Longest Palindromic Substring
 */

// @lc code=start
class Solution {
public:
    string longestPalindrome(string s) {
        if (s.length() == 0) return s;
        int start = 0;
        int res_len = 0;
        for (int i = 0; i < s.length(); i++) {
            int len1 = expandfromCenter(i, i, s);
            int len2 = expandfromCenter(i, i+1, s);
            int len = max(len1, len2);
            if (res_len < len) {
                res_len = len;
                start = (i + len/2)-len+1;  //start = end-len+1
            }
        }
        return s.substr(start, res_len);
    }

    int expandfromCenter(int left, int right, string s) {
        while (left >= 0 && right < s.length()) {
            if (s[left] == s[right]) {
                left--;
                right++;
            } else break;
        }
        return right-left-1;
    }
};
// @lc code=end

File: [5] Longest Palindromic Substring/5.longest-palindromic-substring.java
/*
 * @lc app=leetcode id=5 lang=java
 *
 * [5] Longest Palindromic Substring
 */

// @lc code=start
class Solution {
    public String longestPalindrome(String s) {
        if (s.length() == 1) return s;
        boolean[][] T = new boolean[s.length()][];  // T(i, j): s[i]...s[j] is a palindromic or not
        for (int i = 0; i < s.length(); i++) {
            T[i] = new boolean[s.length()-i];
            T[i][0] = true;
        }
        for (int i = 0; i < s.length()-1; i++) {
            T[i][1] = (s.charAt(i) == s.charAt(i+1));
        }
        for (int k = 2; k < s.length(); k++) {
            for (int i = 0; i < s.length()-k; i++) {
                T[i][k] = (T[i+1][k-2] && (s.charAt(i+k) == s.charAt(i)));
            }
        }
        int ind_i = 0;
        int ind_j = 0;
        int max_len = 0;
        for (int k = 0; k < s.length(); k++) {
            for (int i = 0; i < s.length()-k; i++) {
                if (T[i][k] && max_len < k+1) {
                    ind_i = i;
                    ind_j = i+k;
                    max_len = k+1;
                }
            }
        }
        return s.substring(ind_i, ind_j+1);
    }
}
// @lc code=end

